enhancement(sample transform): add stratified sampling capability

hillmandj opened the pull request:
<!--
**Your PR title must conform to the conventional commit spec!**

  <type>(<scope>)!: <description>

  * `type` = chore, enhancement, feat, fix, docs
  * `!` = OPTIONAL: signals a breaking change
  * `scope` = Optional when `type` is "chore" or "docs", available scopes https://github.com/vectordotdev/vector/blob/master/.github/semantic.yml#L20
  * `description` = short description of the change

Examples:

  * enhancement(file source): Add `sort` option to sort discovered files
  * feat(new source): Initial `statsd` source
  * fix(file source): Fix a bug discovering new files
  * chore(external docs): Clarify `batch_size` option
-->

Resolves [issue](https://github.com/vectordotdev/vector/issues/20921). The goal of this PR is to leverage the use of a HashMap to attempt to enable stratified sampling. The idea is that a user can specify a field in a sampler, such as: `service`, and each event with a unique service will then be sampled together. The motivation behind this is that currently we need to create individual samplers with basically the same configuration over and over in order to achieve the same `rate` for a particular log stream. 

Should the option not be passed, the log event should fallback to the `None` key in the hashmap, and act as a regular counter just as we had before.

A note about this PR/me: I am not a rust developer. This is my first attempt at doing _anything_ with respect to the rust programming language. I am likely doing something wrong (not typing properly, missing edge cases, not testing properly, etc.) and am looking for feedback as to how to make this better / more idiomatic. Hopefully maintainers / actual rust experts will see the motivation and give me some direction in order to make this mergeable. Thanks! 🎉

hillmandj commented on src/transforms/sample/config.rs:
```diff
@@ -44,6 +44,11 @@ pub struct SampleConfig {
 
+    /// A field for grouping sampling by a particular log field. Logs that come in with a unique
+    /// value for the field will be sampled together.
+    #[configurable(metadata(docs::examples = ".log_field"))]
```
probably just want `"log_field"` here without the `.`

hillmandj commented:
@jszwedko I have just accepted the CLA. Please let me know if there's anything else I need to do to get a review on the PR. Thanks!

jszwedko commented:
> @jszwedko I have just accepted the CLA. Please let me know if there's anything else I need to do to get a review on the PR. Thanks!

Thank you! We'll take a look at this shortly.

hillmandj commented:
Tagging @pront as well, hoping to get feedback from ya'll soon!

jszwedko commented on src/transforms/sample/config.rs:
```diff
@@ -44,6 +44,11 @@ pub struct SampleConfig {
+    /// A field for grouping sampling by a particular log field. Logs that come in with a unique
+    /// value for the field will be sampled together.
+    #[configurable(metadata(docs::examples = ".log_field"))]
+    pub group_by: Option<String>,
```
```suggestion
    pub group_by: Option<Template>,
```

Could we make this a template field? That way people could join together multiple fields to create the groupings (e.g. they could do things like `{{ host }}-{{ service }}`. See the `throttle` transform for an example of how the `group_by` would be rendered: https://github.com/vectordotdev/vector/blob/a1edf90e9b78ace1cbb0b579324b1d5b027754b5/src/transforms/throttle.rs#L55-L60

jszwedko commented on src/transforms/sample/transform.rs:
```diff
@@ -8,31 +10,36 @@ use crate::{
+    group_by: Option<String>,
     exclude: Option<Condition>,
-    count: u64,
+    counter: HashMap<String, u64>,
```
```suggestion
    counter: HashMap<Option<String>>, u64>,
```

I'd suggest making the counter hashmap use an `Option` so that, if the user didn't specify a `group_by`, you could use `None` to lookup the counter. This would avoid the need to have a "default" group with a name that might collide.

jszwedko commented on src/transforms/sample/transform.rs:
```diff
@@ -69,13 +80,46 @@ impl FunctionTransform for Sample {
+            default_group_name
+        };
+
+        let counter_value: u64 = match self.counter.entry(counter_key.clone()) {
+            Entry::Occupied(e) => *e.get(),
+            Entry::Vacant(e) => *e.insert(0),
+        };
```
```suggestion
        let counter_value: u64 = match self.counter.entry(counter_key.clone()).or_default();
```

I think this would work?

jszwedko reviewed (commented):
Apologies for the delay in review of this @hillmandj ! This is a valuable enhancement.

I left a few inline comments, but I think this is mostly there. A couple of other notes:

- Could you add a changelog entry? See https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md for a description of the format.
- Could you run `make generate-component-docs`? That should regenerate the documentation with the new field you added.

jszwedko reviewed (approved):
Awesome, thanks for this contribution @hillmandj !

hillmandj commented:
@jszwedko thank you so much for your feedback! I hope that I addressed all of your concerns. Please let me know if you think any more changes are necessary, or if there is some manual testing I should do to ensure this works out before merging. Appreciate your help and happy to contribute! 🎉

maycmlee commented on website/cue/reference/components/transforms/base/sample.cue:
```diff
@@ -6,6 +6,19 @@ base: components: transforms: sample: configuration: {
+		description: """
+			The value to group events into separate buckets to be sampled independently.
+
+			If left unspecified, or if the event doesn't have `group_by`, then the event is not sampled
```
```suggestion
			If left unspecified, or if the event doesn't have `group_by`, then the event is not sampled.
```

maycmlee reviewed (commented):
Just a small edit!

hillmandj commented:
@jszwedko pushed up some small formatting changes after seeing the output of [this](https://github.com/vectordotdev/vector/actions/runs/11352112451/job/31574255444) failed check, and also attempted to address @maycmlee's comment. 

`cargo vdev check fmt` succeeds locally 👍

hillmandj commented:
@jszwedko @pront looks like the PR got removed from the merge queue due to [this](https://github.com/vectordotdev/vector/actions/runs/11354182784/job/31580866222) failing check, which seems unrelated to my change. It appears the job is failing to install the `awscli` ubuntu package. Hoping this is just a random hiccup, but please let me know if there is anything I can do to help! 🙏
